packages/traceability/issuer-api/src/pods/certificate/handlers/certificates-created.handler.ts

```
@@ -0,0 +1,129 @@
+    CertificatePersistedEvent,
+    CertificateUpdatedEvent,
+    SyncCertificateEvent
+} from '../events';
+
+@EventsHandler(CertificatesCreatedEvent)
+export class CertificatesCreatedHandler implements IEventHandler<CertificatesCreatedEvent> {
+    private readonly logger = new Logger(CertificatesCreatedHandler.name);
+
+    constructor(
+        @InjectRepository(Certificate)
+        private readonly repository: Repository<Certificate>,
+        @InjectRepository(UnminedCommitment)
+        private readonly unminedCommitmentRepository: Repository<UnminedCommitment>,
+        private connection: Connection,
+        private readonly blockchainPropertiesService: BlockchainPropertiesService,
+        private readonly eventBus: EventBus
+    ) {}
+
+    async handle({ ids, privateInfo }: CertificatesCreatedEvent): Promise<void> {
+        ids.forEach((id) => this.logger.log(`Detected a new certificate with ID ${id}`));
+
+        const blockchainProperties = await this.blockchainPropertiesService.get();
+
+        const existingCertificates = await this.repository.find({
+            where: {
+                id: In(ids)
+            },
+            relations: ['blockchain']
+        });
+
+        const notExistingCertificates = ids.filter(
+            (id) => !existingCertificates.some((c) => c.id === id)
+        );
+
+        existingCertificates.forEach((c) => {
+            // Re-sync the certificate to sync any eventual changes
+            this.eventBus.publish(new SyncCertificateEvent(c.id));
+        });
+
+        if (notExistingCertificates.length === 0) {
+            return;
+        }
+
+        const newCertificates = await Promise.all(
+            notExistingCertificates.map(async (id) => {
+                try {
+                    return await new CertificateFacade(id, blockchainProperties.wrap()).sync();
+                } catch (e) {
+                    this.logger.error(e.message);
+                    throw e;
```

The only difference is that if any of certificates in "batch" cannot be synced by issuer, then whole process fails, and no certificate is created
Not sure if that is acceptable

---

That's fine, it's better to fail early.